observed_generation.md

```diff
@@ -0,0 +1,74 @@
+used to determine the result of the last reconciliation run. However the
+conditions alone cannot tell which version of the `Spec` was used for the last
+reconciliation, they cannot indicate if there is a newer `Spec` update that has
+not seen by the controller and therefore hasn't reconciled yet.
+
+
+## Solution
+K8s maintains a `Generation` field for each CR. When the `Spec` field of the
+CR is updated the `Generation` of the CR is automatically incremented.
+
+To allow the client to decide if the `Status` field talks about the latest
+version of the `Spec` the `Status` field should tell which version, generation,
+the `Status` is referring to. This is implemented by the
+`Status.ObservedGeneration` field. This field contains the `Generation` value
+of the `Spec` the controller last seen and therefore the `Spec` version the
+`Status` is referring to.
+
+## Usage
+If a client updates the `Spec` field of a CR and wants to decide when such
+`Spec` changes are applied to the running system, then the client needs to look
+at two fields in the CR `Status`:
+* If the `Status.ObservedGeneration == Generation` then the client can be sure
+that the controller seen the last `Spec` version
+* If the `Ready` condition's State is `True` in the `Status.Conditions` list
+then the  client can be sure that the last reconciliation finished
+successfully.
+
+```golang
+if subCR.Generation == subCR.Status.ObservedGeneration &&
+	subCR.Status.Conditions.IsTrue(condition.ReadyCondition) {
+        // subCR is successfully reconciled the latest Spec update
+}
+```
+
+## Implementation
+The `Generation` field is implemented automatically.
+
+The `Status.ObservedGeneration` needs to be implemented by every controller
+individually. This field needs to be bumped to the value of the `Generation`
+field at the start of each `Reconcile` run, after the `Status.Conditions`
+field is reset to signal that a new `Spec` version is being reconciled and
+therefore the old condition values are not valid any more.
+
+## When a CR depends on one or more sub CRs
+We have multiple cases when the Readiness of a parent CR depends on the status
+of other sub CRs (e.g. `Glance` CR `Status` depends on one or more GlanceAPI
+CRs `Status`). If the above implementation is in place then the
+`ObservedGeneration` of the parent CR will not depend on the
+`ObservedGeneration` of the sub CRs. But the Readiness (the Ready=true state)
+of the parent CR depends on the `subCR.Status` fields (mostly the
+`Status.Conditions`) and therefore it needs to consider what `Spec`
+`Generation` those `Status` fields are referring to. So the
+`subCR.Status.ObservedGeneration` needs to be included in the parent CR's
+Readiness calculation. This is needed as the subCR's `Ready` condition alone
+cannot tell if the controller of the subCR see the latest `Spec` update from
+the controller of the parent CR.
+
+## When a CR depends on Pods
+We have CRs that are managing `Pods` via `Deployment`, `StatefulSet` or other
```

Comment:
Cinder and Glance operators are not doing this right.
I'm fixing it in the cinder operator while I add the observedGeneration code.

Reply:
you mean that we should set `ReadyCount` if the `observedGeneration` is the last seen?

Reply:
Yes, because right now this is being done unconditionally (at least in glance and cinder).

Example: https://github.com/openstack-k8s-operators/glance-operator/blob/6b5994d3fe9453fbaad2fff857b152cd75db7d44/controllers/glanceapi_controller.go#L780

Reply:
Yeah, and https://github.com/openstack-k8s-operators/glance-operator/pull/508 cover the glance part of it

Reply:
I did a similar change for manila: https://github.com/openstack-k8s-operators/manila-operator/pull/264/commits/9ec049c88df58489ecc7f913f3f55f0a85a85342